hermes/probe/checknil/check_nil.go

```diff
@@ -0,0 +1,332 @@
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+func verifyNilFile(ctx context.Context, file *nilFile, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	hash := sha1.New()
+	if _, err := io.Copy(hash, file.reader()); err != nil {
+		return probe.NewProbeError(metrics.ProbeFailed, fmt.Errorf("verifyNilFile(): unable to compute checksum of file contents: %w", err))
+	}
+	file.hash = fmt.Sprintf("%x", hash.Sum(nil))
+
+	if !file.validateChecksum() {
+		return probe.NewProbeError(metrics.FileCorrupted, fmt.Errorf("verifyNilFile(): checksum of nil file did not match filename"))
+	}
+
+	journal := &sjpb.StateJournal{}
+	if err := proto.Unmarshal(file.contents, journal); err != nil {
+		return probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("verifyNilFile(): unable to unmarshal nil file proto: %w", err))
+	}
+
+	logger.Infof("verifyNilFile(): unmarshal complete for nil file proto.")
+
+	if err := resolveIntent(ctx, journal, target, client, logger); err != nil {
+		return err
+	}
+
+	if diff := cmp.Diff(target.Journal, journal, protocmp.Transform()); diff != "" {
+		return probe.NewProbeError(metrics.FileCorrupted, fmt.Errorf("verifyNilFile(): journal contents mismatch (-want +got):\n%s", diff))
+	}
+
+	logger.Infof("verifyNilFile(): nil file contents verified against in-memory struct.")
+
+	return probe.NewProbeError(metrics.Success, nil)
+}
+
+// resolveIntent verifies and completes the intent stored in the state journal passed.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- journal: the StateJournal read from the storage system.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+func resolveIntent(ctx context.Context, journal *sjpb.StateJournal, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	if journal.Intent != nil {
+		fileID, err := strconv.Atoi(journal.Intent.Filename[len("Hermes_"):len("Hermes_ID")])
+		if err != nil {
+			return probe.NewProbeError(metrics.ProbeFailed, err)
+		}
+		switch journal.Intent.FileOperation {
+		case sjpb.Intent_CREATE:
+			// TODO(evanSpendlove): Check if the file with this filename has already been created. Should we check the filesystem for the file existing?
+			if journal.Filenames[int32(fileID)] != journal.Intent.Filename {
+				// TODO(evanSpendlove): Call createFile() passing the fileID and return
+				logger.Info("File create not complete yet")
+				journal.Filenames[int32(fileID)] = journal.Intent.Filename
+			}
+		case sjpb.Intent_DELETE:
+			if _, ok := journal.Filenames[int32(fileID)]; ok {
+				// TODO(evanSpendlove): Call  DeleteFile() passing the fileID and return
```

Comment:
What will delete do if you call it with a non-existent file?

Reply:
This should never happen, but if it does, an error is returned.

If Hermes has not been restarted, the in-memory journal should correctly reflect the state of the target system.

If Hermes has been restarted, the in-memory journal will be constructed from the current state of the target system, so this can't be inconsistent unless the file system deletes a file before Hermes reaches this stage, in which case Hermes should record a file missing.